### PR TITLE
implement optimized mapIdx for ND

### PR DESCRIPTION
**Pull request by BenjaminW3**
Based on #323 I implemented optimized versions of the alorithms proposed.
Both algorithms should only need the minimum required number of iterations/operations even for small vectors.
Furthermore, I added a unit test and assertions that we can not work with zero-dimensional vectors.

### CODE REVIEW

**Comment by ax3l** [comment]
Thanks! I rebuild your branch to actually see what you changed compared to my version.

File: include/alpaka/idx/MapIdx.hpp [hunk 1]
```diff
@@ -40,33 +40,35 @@ namespace alpaka
             template<
                 std::size_t TidxDimOut,
                 std::size_t TidxDimIn,
+                bool isSame = ( TidxDimOut == TidxDimIn ),
```
Comment (BenjaminW3) [inline include/alpaka/idx/MapIdx.hpp]: In my version, this line is not required

Comment (ax3l) [inline include/alpaka/idx/MapIdx.hpp]: in that case, my cherry pick was noch successful :-) will fix!

**Comment by ax3l** [comment]
@BenjaminW3 @psychocoderHPC ping :)

**Comment by psychocoderHPC** [comment]
I assigned my self for this pull request because both of you implemented the new version of the mapping.

**Comment by psychocoderHPC** [comment]
thx @ax3l and @BenjaminW3 for this nice pull request

**Pull request merged**